src/lib/Generalize.hs

```diff
@@ -0,0 +1,231 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Generalize (generalizeArgs, generalizeIxDict, Generalized) where
+
+import Control.Monad
+import Control.Category ((>>>))
+
+import Syntax
+import Inference
+import QueryType
+import Name
+import Err
+import MTL1
+import LabeledItems
+
+type Generalized (e::E) (n::S) = (Abs (Nest Binder) e n, [Atom n])
+
+generalizeIxDict :: EnvReader m => Dict n -> m n (Generalized Dict n)
+generalizeIxDict dict = liftGeneralizerM do
+  dict' <- sinkM dict
+  dictTy <- getType dict'
+  dictTyGeneralized <- generalizeType dictTy
+  liftEnvReaderM $ generalizeDict dictTyGeneralized dict'
+{-# INLINE generalizeIxDict #-}
+
+generalizeArgs ::EnvReader m => Nest PiBinder n l -> [Atom n] -> m n (Generalized (ListE Atom) n)
+generalizeArgs reqTys allArgs =  liftGeneralizerM $ runSubstReaderT idSubst do
+  PairE (Abs reqTys' UnitE) (ListE allArgs') <- sinkM $ PairE (Abs reqTys UnitE) (ListE allArgs)
+  ListE <$> go reqTys' allArgs'
+  where
+    go ::Nest PiBinder i i' -> [Atom n] -> SubstReaderT AtomSubstVal GeneralizerM i n [Atom n]
+    go Empty [] = return []
+    go (Nest (PiBinder b ty _) bs) (arg:args) = do
+      ty' <- substM ty
+      arg' <- case ty' of
+        TyKind   -> liftSubstReaderT $ generalizeType arg
+        DictTy _ -> generalizeDict ty' arg
+        _ -> return arg
+      args'' <- extendSubst (b@>SubstVal arg') $ go bs args
+      return $ arg' : args''
+    go _ _ = error "zip error"
+{-# INLINE generalizeArgs #-}
+
+-- === generalizer monad plumbing ===
+
+data GeneralizationEmission n l = GeneralizationEmission (Binder n l) (Atom n)
+type GeneralizationEmissions = RNest GeneralizationEmission
+
+newtype GeneralizerM n a = GeneralizerM {
+  runGeneralizerM' :: DoubleInplaceT Env GeneralizationEmissions UnitB HardFailM n a }
+  deriving (Functor, Applicative, Monad, MonadFail, Fallible, ScopeReader, EnvReader, EnvExtender)
+
+liftGeneralizerM
+  :: EnvReader m
+  => (forall l. DExt n l => GeneralizerM l (e l))
+  -> m n (Generalized e n)
+liftGeneralizerM cont = do
+  env <- unsafeGetEnv
+  Distinct <- getDistinct
+  Abs emissions (DoubleInplaceTResult UnitB e) <- return $ runHardFail $
+    runDoubleInplaceT env $ runGeneralizerM' cont
+  let (bs, vals) = hoistGeneralizationVals (unRNest emissions)
+  return (Abs bs e, vals)
+  where
+    -- TODO: something not O(N^2)
+    hoistGeneralizationVals :: Nest GeneralizationEmission n l -> (Nest Binder n l, [Atom n])
+    hoistGeneralizationVals Empty = (Empty, [])
+    hoistGeneralizationVals (Nest (GeneralizationEmission b val) bs) = do
+      let (bs', vals) = hoistGeneralizationVals bs
+      case hoist b (ListE vals) of
+        HoistSuccess (ListE vals') -> (Nest b bs', val:vals')
+        HoistFailure _ -> error "should't happen"
```

Comment:
Why can't this happen? It's not immediately obvious. Perhaps add a short comment too?